cli/command/stack/kubernetes/services.go

```diff
@@ -2,43 +2,136 @@ package kubernetes
 
 	"github.com/docker/cli/cli/command/formatter"
 	"github.com/docker/cli/cli/command/stack/options"
 	"github.com/docker/cli/kubernetes/labels"
+	"github.com/docker/docker/api/types/filters"
+	appsv1beta2 "k8s.io/api/apps/v1beta2"
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+var supportedServicesFilters = map[string]bool{
+	"mode":  true,
+	"name":  true,
+	"label": true,
+}
+
+func generateValuedLabelsSelector(valuedLabels map[string][]string) []string {
+	var result []string
+	for k, v := range valuedLabels {
+		if len(v) == 1 {
+			result = append(result, fmt.Sprintf("%s=%s", k, v[0]))
+		} else {
+			sort.Strings(v)
+			result = append(result, fmt.Sprintf("%s in (%s)", k, strings.Join(v, ",")))
+		}
+	}
+	return result
+}
+
+func parseLabelFilters(rawFilters []string) ([]string, map[string][]string) {
+	var presentLabels []string
+	valuedLabels := map[string][]string{}
+	for _, rawLabel := range rawFilters {
+		equalIndex := strings.Index(rawLabel, "=")
+		if equalIndex == -1 {
+			presentLabels = append(presentLabels, rawLabel)
+		} else {
+			key := rawLabel[:equalIndex]
+			val := rawLabel[equalIndex+1:]
+			valuedLabels[key] = append(valuedLabels[key], val)
+		}
+	}
+	return presentLabels, valuedLabels
+}
+
+func generateLabelSelector(f filters.Args, stackName string) string {
+	names := f.Get("name")
+	sort.Strings(names)
+
+	for _, n := range names {
+		if strings.HasPrefix(n, stackName+"_") {
+			// also accepts with unprefixed service name (for compat with existing swarm scripts where service names are prefixed by stack names)
+			names = append(names, strings.TrimPrefix(n, stackName+"_"))
+		}
+	}
+
+	presentLabels, valuedLabels := parseLabelFilters(f.Get("label"))
+	selectors := append(presentLabels, labels.SelectorForStack(stackName, names...))
+	selectors = append(selectors, generateValuedLabelsSelector(valuedLabels)...)
+	return strings.Join(selectors, ",")
+}
+
+func getResourcesForServiceList(dockerCli *KubeCli, filters filters.Args, labelSelector string) (*appsv1beta2.ReplicaSetList, *appsv1beta2.DaemonSetList, *corev1.ServiceList, error) {
+	client, err := dockerCli.composeClient()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	replicas := client.ReplicaSets()
+	daemons := client.DaemonSets()
+	modes := filters.Get("mode")
+
+	var replicasList *appsv1beta2.ReplicaSetList
+	if len(modes) == 0 || filters.ExactMatch("mode", "replicated") {
+		if replicasList, err = replicas.List(metav1.ListOptions{LabelSelector: labelSelector}); err != nil {
+			return nil, nil, nil, err
+		}
+	} else {
+		replicasList = &appsv1beta2.ReplicaSetList{}
+	}
+
+	var daemonsList *appsv1beta2.DaemonSetList
+	if len(modes) == 0 || filters.ExactMatch("mode", "global") {
+		if daemonsList, err = daemons.List(metav1.ListOptions{LabelSelector: labelSelector}); err != nil {
+			return nil, nil, nil, err
+		}
+	} else {
+		daemonsList = &appsv1beta2.DaemonSetList{}
```

Comment:
Is this needed (or done already by `var replicasList *appsv1beta2.DaemonSetList ` ?)